core/client/fs/src/main/java/alluxio/client/file/dora/netty/NettyDataReader.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file.dora.netty;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.client.file.dora.DoraDataReader;
+import alluxio.client.file.dora.netty.PartialReadException.CauseType;
+import alluxio.proto.dataserver.Protocol;
+import alluxio.wire.WorkerNetAddress;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.channels.WritableByteChannel;
+import java.util.function.Supplier;
+
+/**
+ * Positioned Netty data reader.
+ */
+public class NettyDataReader implements DoraDataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(NettyDataReader.class);
+  private final FileSystemContext mContext;
+  private final WorkerNetAddress mAddress;
+  private final Supplier<Protocol.ReadRequest.Builder> mRequestBuilder;
+  private boolean mClosed = false;
+
+  /**
+   * Constructor.
+   *
+   * @param context
+   * @param address
+   * @param requestBuilder
+   */
+  public NettyDataReader(FileSystemContext context, WorkerNetAddress address,
+      Protocol.ReadRequest.Builder requestBuilder) {
+    mContext = context;
+    mAddress = address;
+    // clone the builder so that the initial values does not get overridden
+    mRequestBuilder = requestBuilder::clone;
+  }
+
+  @Override
+  public int read(long offset, WritableByteChannel outChannel, int length)
+      throws PartialReadException {
+    Preconditions.checkState(!mClosed, "Reader is closed");
+    Protocol.ReadRequest.Builder builder = mRequestBuilder.get()
+        .setLength(length)
+        .setOffset(offset)
+        .clearCancel();
+    NettyClientStateMachine clientStateMachine =
+        new NettyClientStateMachine(mContext, mAddress, builder, outChannel);
+    clientStateMachine.run();
+    int bytesRead = clientStateMachine.getBytesRead();
+    PartialReadException exception = clientStateMachine.getException();
```

Comment:
The `NettyClientStateMachine` can only do `read()` operation.
It is not a general purpose Netty client.
In future, is it possible to do `getStatus()`? and it is easy to add handlers for other request/reply?

Reply:
Sorry if the naming is misleading. This is part of the data reader implementation, so it only handles data I/O. All metadata related operations should continue to go through gRPC as they do not need streaming and is generally less performance sensitive. gRPC does a good job in handling unary calls and smaller requests/responses.

I think to avoid confusion, I will rename this to `NettyDataReaderStateMachine`. How does it sound to you?

Reply:
Yeah. That's what I meant.

Before reviewing this PR, I was thinking we would implement this with gRPC unary request/reply, as you mentioned, gRPC can handle small unary request/reply well.
That means if this is a small and positioned read, we will do the read in gRPC handler.

But no objections to implement it as this PR does.